crates/datasources/src/bson/builder.rs

```diff
@@ -102,10 +100,48 @@ impl RecordStructBuilder {
         // Append nulls to all columns not included in the doc.
         for (idx, did_set) in cols_set.iter().enumerate() {
             if !did_set {
-                // Add nulls...
-                let typ = self.fields.get(idx).unwrap().data_type(); // Programmer error if data type doesn't exist.
-                let col = self.builders.get_mut(idx).unwrap(); // Programmer error if column doesn't exist.
-                append_null(typ, col.as_mut())?;
+                // Add null...
+                self.add_value_at_index(idx, None)?;
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn project_and_append(&mut self, doc: &RawDocument) -> Result<()> {
+        let mut cols_set: BitVec<u8, Lsb0> = BitVec::repeat(false, self.fields.len());
+
+        for iter_result in doc {
+            match iter_result {
+                Ok((key, val)) => {
+                    if let Some(idx) = self.field_index.get(key) {
+                        let idx = idx.to_owned();
```

Comment:
```suggestion
                    if let Some(&idx) = self.field_index.get(key) {
```

Since it's a pattern.

Also for `&usize` (and `&i64`, etc) you can dereference with `*idx` instead of having to do `let idx = idx.to_owned()`.